summaries/search-based-survey.md

```diff
@@ -0,0 +1,335 @@
+
+Backwards path traversal has an advantage over forward traversal in the sense that no storage is required for intermediate symbolic expressions of variables. Forward traversal however allows for early detection of infeasible paths.
+
+Constraint satisfaction problems are NP-complete. However, combinations of heuristics and linear programming techniques can be applied to lighten the process.
+
+Loops are executed *K* times, where *K* is specified by the tester or system.
+
+Procedure calls can be inlined as a solution, however the number of paths can grow rapidly with this approach
+
+#### Domain Reduction
+*Constraint-based Testing* builds up a constraint system which describes the test goal. 
+*Reachability constraints* withing the constraint system describe the conditions required for the reachability of statements.
+*Necessity constraints* describe the conditions that will kill a mutant.
+
+Using Symbolic Execution we can develop constraints in terms of input variables and perform *Domain Reduction*  to attempt a solution to the constraints. 
+> Large search space ---constraints---> Reduction of search space
+
+#### Dynamic Domain Reduction
+Dynamic Domain Reduction is a *static analysis technique* that starts with domains of the input variables and reduce these dynamically during the symbolic execution stage in contrast to normal Domain Reduction. 
+
+*This technique still suffers from difficulties with computed storage locations, loops and non-ordinal variable types(e.g. Enums).*
+
+### Dynamic Structural Test Data Generation
+*Dynamic methods* use some input to execute a program and uses instrumentation to observer the results.
+Many of the issues regarding Symbolic Execution are resolved by directly resolving e.g. pointer values and array subscripts.
+
+#### Random Testing
+Executes the program with random inputs and observes executed program structures.
+When using this technique, code that has low probability of being executed is often not covered, in such cases more directed search techniques are required.
+
+##### Applying Local Search
+In the Local Search technique, the tester selects a path through the program and produces a straight-line version of that path.
+Path constraints are constructed with constant for each constraint estimating the satisfiability of the constraint.
+A function *f* is constructed out of all of these constraints.
+*f* provides an estimate of the satisfiability of all constraints.
+Using numerical maximisation techniques, the input values are maximized to satisfy as many constraints as possible.
+One major draw-back of this was that that run-time errors could occur, since the dependency between constraints were not taken into account, which could create sequences, which in practice are impossible to execute.
```

Comment:
Google says "drawback", the document used "draw-back". I'm going with Google on this one.